### PR TITLE
feat(llm): optional Claude subscription provider via Claude Code

### PR DESCRIPTION
**Pull request by debsahu**
## Summary

Adds an opt-in `claude-code` provider so a single-user, self-hosted Mike can run Claude on the operator's Claude Pro or Max subscription through the Claude Agent SDK, without an Anthropic API key.

## Why / Motivation

Some self-hosters, including solo practitioners reviewing their own documents, already pay for a Claude subscription and would rather not also pay per token. Anthropic permits Claude Code (`claude -p` and the Agent SDK) for this kind of local, personal use. The provider is disabled by default and does not change behavior for existing deployments.

I know CONTRIBUTING asks for a PRD issue before changes that span several subsystems. I built this for my own setup first and am opening the PR so the working code is available for discussion. I am happy to open the PRD issue, split this into smaller PRs, or rework the approach if you would prefer something different.

## Changes

- **Backend provider.** New `backend/src/lib/llm/claudeCode.ts`, reached from `streamWithProvider` and `completeWithProvider`. Model ids are `claude-code/opus`, `claude-code/sonnet` and `claude-code/haiku`.
- **Tools.** Mike's existing tool schemas are served to Claude Code by an in-process MCP server. Each tool call goes through the existing `runTools` path, reusing `ToolExecutionBatcher`, so the document tools, CourtListener, workflows, Word client tools and `ask_inputs` behave as they do with other providers.
- **Streaming.** Text and thinking deltas map to the existing callbacks. Reasoning levels map to Claude's `effort` setting.
- **Isolation.** Claude Code runs with built-in tools disabled, no setting sources, strict MCP config, no session persistence and an empty working directory. The subprocess receives an allowlist of environment variables, so Mike's database, storage and provider secrets are not passed to it.
- **Gating.** `CLAUDE_CODE_ENABLED` controls availability. `CLAUDE_CODE_OAUTH_TOKEN` comes from `claude setup-token`. `hasApiKeyForModel` and the tabular key check treat the provider as keyless only while it is enabled.
- **Model listing.** `GET /models/claude-code` returns the models only when enabled. The web model picker, the settings pickers and the Word add-in list them the same way they list Ollama models. The Ollama polling hook was generalized into `dynamicModelStore.ts` so both hooks share it.
- **Docs.** New `docs/claude-subscription.md`, linked from the README and the docs index. `backend/.env.example` is updated.

## Tradeoffs & risks

- **Dependency.** Adds `@anthropic-ai/claude-agent-sdk`, which bundles a platform-specific Claude Code binary (about 220 MB on linux-arm64) as an optional dependency. The image builds without Dockerfile changes, but it is larger.
- **Latency.** Each request starts a `claude` process, which adds about 1 second before the first token.
- **History.** Prior turns are sent as a delimited text transcript in one prompt. This matches the text-only history the other providers already receive.
- **Scope of use.** The token belongs to one personal account. The docs state this is for single-user local deployments and should not be enabled on shared servers.
- **Out of scope.** No schema or migration changes, and no change to existing providers.

## How verified

- `npm test --prefix backend`: 130 files passed, 1600 tests passed. This includes the new `src/lib/__tests__/claudeCode.test.ts` (11 tests: stream mapping, tool routing with the model's tool ids, error and abort propagation, the max-turns partial result, the disabled gate, and `completeText`).
- `npx tsc --noEmit -p backend`: passes.
- Frontend: `npx tsc --noEmit`, `npm run lint` (no new warnings) and `npx vitest run` (156 files, 1076 tests) pass.
- Word add-in: `npm run typecheck` passes.
- Manual, against a real Max subscription, on the host and inside the backend container started with `docker compose up --build`:
  - Title generation with `claude-code/haiku`.
  - A chat turn with `claude-code/sonnet` that called a tool and answered from the tool result.
  - An `ask_inputs` style pause, which stops the query and rethrows the pause error.
  - A full chat in the web UI with `claude-code/opus` that used `read_document` on an uploaded DOCX.

## Checklist

- [x] Ran the relevant build/test command for the area changed.
- [x] Reviewed `git diff` and removed unrelated changes.
- [x] Updated docs / env examples if setup, config, or behavior changed.
- [x] No secrets, API keys, real documents, or `.env` files committed.

### CODE REVIEW

File: backend/src/lib/llm/claudeCode.ts [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    } catch (error) {
+      toolFailure ??= error;
+      abortController.abort(error);
+      throw error;
+    }
+  };
+
+  let fullText = "";
+  let thinkingOpen = false;
+  const q = loaded.query({
```
Comment (DIYA73) [inline backend/src/lib/llm/claudeCode.ts:277]: loaded.query() passes abortController (line 282), which handles client disconnect correctly. One thing worth considering: is there a wall-clock timeout for the case where Claude Code is running but not making progress — for example if it loops on a tool call within maxTurns? The maxTurns bound covers iteration count but not elapsed time. A slow or stuck turn could hold the SSE connection open indefinitely. Would an optional timeoutMs parameter make sense here, similar to how other streaming paths guard against hung connections?

Comment (debsahu) [reply backend/src/lib/llm/claudeCode.ts:364]: Added an idle guard rather than a wall-clock one.

The API providers stream over HTTP, so a wedged request eventually dies at the socket. This provider drives a local `claude` subprocess, which has nothing underneath it to time out, so the case you describe would hold the SSE connection until the client disconnects. Closest existing precedent is `uploadConversionTimeoutMs`, which bounds a wedged `soffice` child for the same reason.

The timer measures silence rather than total elapsed time. A wall-clock cap would need a generous default to avoid killing a legitimately long multi-tool turn, which weakens the guard; an idle timer targets the stuck case directly. It also pauses while one of Mike's own tools is running, since Claude Code waiting on our tool result is not a stall on its side.

`CLAUDE_CODE_IDLE_TIMEOUT_MS`, default 120s, clamped to 10s-30min, `0` disables. Applied to `completeText` too, since title generation is awaited before the stream closes. Three tests cover the abort, the slow-tool false positive, and the disable switch.

File: docs/claude-subscription.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  `backend/src/lib/llm/claudeCode.ts`, behind the same `streamChatWithTools` /
+  `completeText` interface as every other provider.
+- **Tools stay in Mike:** Mike's tools are passed to Claude Code through an
+  in-process MCP server, and each call runs through Mike's normal tool
+  dispatcher.
+- **Isolation:** Claude Code runs with its built-in tools (Bash, file access,
+  web) disabled, no settings, hooks, plugins or `CLAUDE.md` files, and an empty
+  working directory. It can use only Mike's tools.
+- **Minimal environment:** the subprocess gets a small allowlist of environment
+  variables. Mike's database, storage and provider secrets are not passed to it.
```
[the remark this answers is not in the document]
Comment (debsahu) [reply docs/claude-subscription.md]: Fair, the wording was ambiguous. Rewritten to say the token is forwarded because it is what authenticates the subscription, and that what stays out is Mike's own database, storage and other-provider keys. Also called out `ANTHROPIC_API_KEY` specifically, since an inherited one would make Claude Code bill the key instead of the subscription.

File: frontend/src/app/hooks/dynamicModelStore.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    useModels: () => T[];
+} {
+    let cache: T[] | null = null;
+    let inflight: Promise<T[]> | null = null;
+    const listeners = new Set<() => void>();
+
+    function load(force = false): Promise<T[]> {
+        if (force) {
+            cache = null;
+            inflight = null;
```
[the remark this answers is not in the document]
Comment (debsahu) [reply frontend/src/app/hooks/dynamicModelStore.ts]: The race is real. A superseded request's success handler can commit a stale list over a fresher one, and its failure handler clears the newer request's `inflight`.

Removed `refresh()` rather than adding a generation counter, because nothing calls it. `refreshClaudeCodeModels` and `refreshOllamaModels` were both exported and never used, and the Ollama one was already unused before this branch. The catalog is fixed for the life of the page, so a picker that needs new models reloads. If a caller turns up later, it should come back with the generation guard and a test.

**Comment by debsahu** [comment]
On the out-of-band note about `fetchWithRefresh` not retrying after a 401: leaving that as is.

`refreshSession()` does not mint a new session. It re-reads `GET /auth/session` with the same cookies and updates the client-side login gate. The backend already refreshes HttpOnly sessions before handlers run, so a 401 means the cookie can no longer be refreshed and a reissued request would get another 401. The comment above the function says this. It is also pre-existing code this PR does not touch.

File: backend/src/lib/llm/models.ts [hunk 4]
```diff
@@ -52,6 +52,30 @@ export const GEMINI_LOW_MODELS = [
...
+export const CLAUDE_CODE_LOW_MODELS = ["claude-code/haiku"] as const;
+export const CLAUDE_CODE_MODELS = [
+    ...CLAUDE_CODE_MAIN_MODELS,
+    ...CLAUDE_CODE_LOW_MODELS,
+] as const;
+
+export function isClaudeCodeEnabled(): boolean {
+    return /^(1|true|yes)$/i.test(
+        process.env.CLAUDE_CODE_ENABLED?.trim() ?? "",
+    );
```
[the remark this answers is not in the document]
Comment (debsahu) [reply backend/src/lib/llm/models.ts:72]: Leaving this as is.

The premise is right: `isClaudeCodeEnabled()` reads only `CLAUDE_CODE_ENABLED`, so an operator who enables it without a token or a `claude` login gets three models that fail at first use.

But the fix is not a quick win. You correctly say not to require `CLAUDE_CODE_OAUTH_TOKEN`, since an existing login is equally valid, which leaves probing that login. That means either spawning the CLI on every catalog request, which the PR already measures at about a second, or reading `~/.claude` credentials, whose format is internal and would break on us without warning.

Against that, the current failure is loud, immediate, and already covered in the troubleshooting section. A misconfigured single-user self-host getting a clear error the first time it runs seems better than a background probe that is either slow or brittle.

Happy to revisit if you would rather have a cached readiness check, but it is a bigger change than this PR should carry.

File: frontend/src/app/lib/modelAvailability.ts [hunk 5]
```diff
@@ -39,6 +42,7 @@ export function isProviderAvailable(
     apiKeys: ApiKeyState,
 ): boolean {
     if (provider === "ollama") return true; // local, no key needed
+    if (provider === "claude-code") return true; // subscription, no key needed
```
[the remark this answers is not in the document]
Comment (debsahu) [reply frontend/src/app/lib/modelAvailability.ts:45]: Leaving this as is, but it did turn up a real bug next to it.

On the finding itself: this is not specific to Claude Code. In all three places you flag, `claude-code/` sits on the line immediately after `ollama/` and is treated identically, and Ollama has behaved this way since well before this branch:

```
id.startsWith("ollama/") ||
id.startsWith("claude-code/") ||
```

The add-in states the policy deliberately: "fails OPEN: the backend authoritatively rejects a model it cannot serve, so blocking sends here on a flaky WKWebView request would brick the composer." And the backend does reject it, in `resolveEffectiveChatModel` via `hasApiKeyForModel`, before the provider is reached. Gating claude-code client-side but not Ollama would make the two runtime-discovered providers inconsistent and reverse that decision, which is not something this PR should decide.

What checking this did surface is that the backend rejection said the wrong thing. Chat answered 422 `missing_api_key` with "An API key is required to use claude-code/opus", and tabular used the right sentence under that same code, which `TabularReviewView` turns into the ApiKeyMissingPopup, asking for a key that does not exist for a subscription provider. Both now answer 400 `model_unavailable` with one shared sentence, and `missingModelApiKey` is left to mean only a key the user could actually add. Fixed in 0e06ecb with regression tests on both paths.

**Comment by debsahu** [comment]
On the out-of-band note about `word-addin/src/taskpane/lib/modelCatalog.ts` requiring `claude-code/*` IDs to be present in the fetched catalog: leaving that as is, for the same reason as the web one.

`isModelAvailable` there treats `claude-code/` exactly as it treats `ollama/`, on the adjacent line, and the comment a few lines below sets out the policy on purpose:

> fails OPEN: the backend authoritatively rejects a model it cannot serve, so blocking sends here on a flaky WKWebView request would brick the composer for requests the backend would happily accept.

Requiring catalog presence for Claude Code but not Ollama would make the add-in inconsistent with itself and with the web app, and would reverse a documented decision about WKWebView reliability. The backend rejection is the authority, and this PR has just corrected what that rejection says.

File: backend/src/lib/llm/claudeCode.ts (hunk 1 above)
Comment (debsahu) [reply backend/src/lib/llm/claudeCode.ts:364]: Added the wall-clock bound you asked for, in 704d583, so both now apply.

`CLAUDE_CODE_TIMEOUT_MS` caps the whole turn regardless of how busy it looks. It never pauses and never resets, which is the part the idle timer genuinely cannot do: the idle timer restarts on every byte, so a turn that keeps emitting output or tool calls without converging resets it forever. You were right that `maxTurns` bounds iterations rather than elapsed time, and this closes that gap.

Default is ten minutes, clamped to 60s-4h, `0` disables. It is deliberately generous because, unlike the idle timer, a total cap cannot tell productive work from a loop, so it reads better as a backstop than as the first line of defence. The idle guard at 120s still does the early catching, and the error names whichever bound was reached. Both apply to `completeText` too.

Tests cover a turn that streams continuously and never finishes, one that finishes just inside the budget, and the disable switch. Each guard's tests disable the other so they cannot mask one another, which the first version of them did.

File: backend/src/lib/llm/claudeCode.ts [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+    },
+  });
+  watchdog.touch();
+  try {
+    for await (const message of q) {
+      watchdog.touch();
+      if (message.type !== "result") continue;
+      const error = resultError(message);
+      if (error) throw error;
+      return message.subtype === "success" ? message.result : "";
```
[the remark this answers is not in the document]
Comment (debsahu) [reply backend/src/lib/llm/claudeCode.ts:499]: Good catch, applied in c6bd011 with your placement.

The `return` inside the loop reaches `finally` and skips the checks after it, so a result that arrived as a timer fired was reported as a success and the bound silently did not hold. Putting the checks ahead of `resultError` also keeps the message about the bound that was reached rather than the abort it caused, which is what the SDK reports once the controller fires.

`streamClaudeCode` has no early return, so it already fell through to those checks and is unaffected.

Covered by a table test over both bounds: the timer fires, then an already-queued success result arrives, and the call must reject rather than return the text. Removing the two lines fails both cases.

[the remark this answers is not in the document]
Comment (DIYA73) [reply backend/src/lib/llm/claudeCode.ts:499]: Thanks for the detailed explanation and for applying the fix. The table test covering both bounds looks solid — the edge case around a queued success arriving after the timer fires is exactly what I was worried about. LGTM on that part.
